File: go/bcenvelope/elide.go
package bcenvelope

import (
	bccomponents "github.com/nickel-blockchaincommons/bccomponents-go"
)

// ObscureType identifies the type of obscuration applied to an element.
type ObscureType int

const (
	ObscureTypeElided     ObscureType = iota
	ObscureTypeEncrypted
	ObscureTypeCompressed
)

// ObscureAction specifies what to do when obscuring an element.
type ObscureAction int

const (
	ObscureActionElide    ObscureAction = iota
	ObscureActionCompress
	ObscureActionEncrypt
)

// ObscureActionEncryptWithKey holds a symmetric key for encryption-based obscuration.
type ObscureActionEncryptWithKey struct {
	Key bccomponents.SymmetricKey
}

// --- Elision ---

// Elide returns the elided variant of this envelope (only its digest remains).
func (e *Envelope) Elide() *Envelope {
	if _, ok := e.envelopeCase.(*ElidedCase); ok {
		return e
	}
	return newElided(e.Digest())
}

// --- Set-based elision ---

// ElideRemovingSet elides elements whose digests are in the target set.
func (e *Envelope) ElideRemovingSet(target map[bccomponents.Digest]struct{}) *Envelope {
	return e.ElideSet(target, false)
}

// ElideRevealingSet elides elements whose digests are not in the target set.
func (e *Envelope) ElideRevealingSet(target map[bccomponents.Digest]struct{}) *Envelope {
	return e.ElideSet(target, true)
}

// ElideSet elides elements based on the target set and revealing flag.
func (e *Envelope) ElideSet(target map[bccomponents.Digest]struct{}, isRevealing bool) *Envelope {
	return e.ElideSetWithAction(target, isRevealing, ObscureActionElide, nil)
}

// ElideSetWithAction applies obscuration based on the target set using the given action.
func (e *Envelope) ElideSetWithAction(
	target map[bccomponents.Digest]struct{},
	isRevealing bool,
	action ObscureAction,
	encryptKey *ObscureActionEncryptWithKey,
) *Envelope {
	selfDigest := e.Digest()
	_, inTarget := target[selfDigest]
	shouldObscure := inTarget != isRevealing

	if shouldObscure {
		switch action {
		case ObscureActionElide:
			return e.Elide()
		case ObscureActionCompress:
			compressed, err := e.Compress()
			if err != nil {
				return e.Elide()
			}
			return compressed
		case ObscureActionEncrypt:
			if encryptKey != nil {
				data := e.TaggedCBOR().ToCBORData()
				msg := encryptKey.Key.EncryptWithDigest(data, selfDigest, nil)
				env, err := newWithEncrypted(&msg)
				if err != nil {
					return e.Elide()
				}
				return env
			}
			return e.Elide()
		default:
			return e.Elide()
		}
	}

	// Not obscuring this node — recurse into children
	switch c := e.envelopeCase.(type) {
	case *AssertionCase:
		pred := c.Assertion.Predicate().ElideSetWithAction(target, isRevealing, action, encryptKey)
		obj := c.Assertion.Object().ElideSetWithAction(target, isRevealing, action, encryptKey)
		elidedAssertion := NewAssertion(
			EnvelopeEncodableEnvelope{pred},
			EnvelopeEncodableEnvelope{obj},
		)
		return newWithAssertion(elidedAssertion)
	case *NodeCase:
		elidedSubject := c.Subject.ElideSetWithAction(target, isRevealing, action, encryptKey)
		elidedAssertions := make([]*Envelope, len(c.Assertions))
		for i, a := range c.Assertions {
			elidedAssertions[i] = a.ElideSetWithAction(target, isRevealing, action, encryptKey)
		}
		return newWithUncheckedAssertions(elidedSubject, elidedAssertions)
	case *WrappedCase:
		elidedEnvelope := c.Envelope.ElideSetWithAction(target, isRevealing, action, encryptKey)
		return newWrapped(elidedEnvelope)
	default:
		return e
	}
}

// --- Target convenience methods ---

// ElideRemovingTarget elides the single target element.
func (e *Envelope) ElideRemovingTarget(target DigestProvider) *Envelope {
	return e.ElideRemovingSet(digestSet(target.Digest()))
}

// ElideRevealingTarget reveals only the single target element, eliding everything else.
func (e *Envelope) ElideRevealingTarget(target DigestProvider) *Envelope {
	return e.ElideRevealingSet(digestSet(target.Digest()))
}

// ElideRemovingArray elides elements whose digests are in the targets slice.
func (e *Envelope) ElideRemovingArray(targets []DigestProvider) *Envelope {
	return e.ElideRemovingSet(digestSetFromArray(targets))
}

// ElideRevealingArray reveals only the target elements, eliding everything else.
func (e *Envelope) ElideRevealingArray(targets []DigestProvider) *Envelope {
	return e.ElideRevealingSet(digestSetFromArray(targets))
}

// ElideArrayWithAction obscures array elements using the given action and reveal direction.
func (e *Envelope) ElideArrayWithAction(
	targets []DigestProvider,
	isRevealing bool,
	action ObscureAction,
	encryptKey *ObscureActionEncryptWithKey,
) *Envelope {
	return e.ElideSetWithAction(digestSetFromArray(targets), isRevealing, action, encryptKey)
}

// ElideArray elides array elements using the given reveal direction.
func (e *Envelope) ElideArray(targets []DigestProvider, isRevealing bool) *Envelope {
	return e.ElideArrayWithAction(targets, isRevealing, ObscureActionElide, nil)
}

// ElideTargetWithAction obscures a single target using the given action and reveal direction.
func (e *Envelope) ElideTargetWithAction(
	target DigestProvider,
	isRevealing bool,
	action ObscureAction,
	encryptKey *ObscureActionEncryptWithKey,
) *Envelope {
	return e.ElideArrayWithAction([]DigestProvider{target}, isRevealing, action, encryptKey)
}

// ElideTarget elides a single target using the given reveal direction.
func (e *Envelope) ElideTarget(target DigestProvider, isRevealing bool) *Envelope {
	return e.ElideTargetWithAction(target, isRevealing, ObscureActionElide, nil)
}

// ElideRemovingArrayWithAction removes array elements using the given action.
func (e *Envelope) ElideRemovingArrayWithAction(
	targets []DigestProvider,
	action ObscureAction,
	encryptKey *ObscureActionEncryptWithKey,
) *Envelope {
	return e.ElideArrayWithAction(targets, false, action, encryptKey)
}

// ElideRevealingArrayWithAction reveals array elements, obscuring everything else.
func (e *Envelope) ElideRevealingArrayWithAction(
	targets []DigestProvider,
	action ObscureAction,
	encryptKey *ObscureActionEncryptWithKey,
) *Envelope {
	return e.ElideArrayWithAction(targets, true, action, encryptKey)
}

// ElideRemovingTargetWithAction removes a single target using the given action.
func (e *Envelope) ElideRemovingTargetWithAction(
	target DigestProvider,
	action ObscureAction,
	encryptKey *ObscureActionEncryptWithKey,
) *Envelope {
	return e.ElideTargetWithAction(target, false, action, encryptKey)
}

// ElideRevealingTargetWithAction reveals a single target, obscuring everything else.
func (e *Envelope) ElideRevealingTargetWithAction(
	target DigestProvider,
	action ObscureAction,
	encryptKey *ObscureActionEncryptWithKey,
) *Envelope {
	return e.ElideTargetWithAction(target, true, action, encryptKey)
}

// --- Unelide ---

// Unelide replaces this elided envelope with the provided envelope if their digests match.
func (e *Envelope) Unelide(envelope *Envelope) (*Envelope, error) {
	if e.Digest().Equal(envelope.Digest()) {
		return envelope, nil
	}
	return nil, ErrInvalidDigest
}

// --- Walk-based operations ---

// WalkUnelide recursively walks the envelope and replaces elided nodes with
// matching envelopes from the provided slice.
func (e *Envelope) WalkUnelide(envelopes []*Envelope) *Envelope {
	envelopeMap := make(map[bccomponents.Digest]*Envelope)
	for _, env := range envelopes {
		envelopeMap[env.Digest()] = env
	}
	return e.walkUnelideWithMap(envelopeMap)
}

func (e *Envelope) walkUnelideWithMap(envelopeMap map[bccomponents.Digest]*Envelope) *Envelope {
	switch c := e.envelopeCase.(type) {
	case *ElidedCase:
		if replacement, ok := envelopeMap[e.Digest()]; ok {
			return replacement
		}
		return e
	case *NodeCase:
		newSubject := c.Subject.walkUnelideWithMap(envelopeMap)
		newAssertions := make([]*Envelope, len(c.Assertions))
		changed := !newSubject.IsIdenticalTo(c.Subject)
		for i, a := range c.Assertions {
			newAssertions[i] = a.walkUnelideWithMap(envelopeMap)
			if !newAssertions[i].IsIdenticalTo(a) {
				changed = true
			}
		}
		if !changed {
			return e
		}
		return newWithUncheckedAssertions(newSubject, newAssertions)
	case *WrappedCase:
		newEnv := c.Envelope.walkUnelideWithMap(envelopeMap)
		if newEnv.IsIdenticalTo(c.Envelope) {
			return e
		}
		return newEnv.Wrap()
	case *AssertionCase:
		newPred := c.Assertion.Predicate().walkUnelideWithMap(envelopeMap)
		newObj := c.Assertion.Object().walkUnelideWithMap(envelopeMap)
		if newPred.IsIdenticalTo(c.Assertion.Predicate()) && newObj.IsIdenticalTo(c.Assertion.Object()) {
			return e
		}
		return NewAssertionEnvelope(EnvelopeEncodableEnvelope{newPred}, EnvelopeEncodableEnvelope{newObj})
	default:
		return e
	}
}

// WalkReplace recursively walks the envelope and replaces nodes whose digests
// are in the target set with the replacement envelope.
func (e *Envelope) WalkReplace(target map[bccomponents.Digest]struct{}, replacement *Envelope) (*Envelope, error) {
	if _, ok := target[e.Digest()]; ok {
		return replacement, nil
	}
	switch c := e.envelopeCase.(type) {
	case *NodeCase:
		newSubject, err := c.Subject.WalkReplace(target, replacement)
		if err != nil {
			return nil, err
		}
		newAssertions := make([]*Envelope, len(c.Assertions))
		changed := !newSubject.IsIdenticalTo(c.Subject)
		for i, a := range c.Assertions {
			newAssertions[i], err = a.WalkReplace(target, replacement)
			if err != nil {
				return nil, err
			}
			if !newAssertions[i].IsIdenticalTo(a) {
				changed = true
			}
		}
		if !changed {
			return e, nil
		}
		return newWithAssertions(newSubject, newAssertions)
	case *WrappedCase:
		newEnv, err := c.Envelope.WalkReplace(target, replacement)
		if err != nil {
			return nil, err
		}
		if newEnv.IsIdenticalTo(c.Envelope) {
			return e, nil
		}
		return newEnv.Wrap(), nil
	case *AssertionCase:
		newPred, err := c.Assertion.Predicate().WalkReplace(target, replacement)
		if err != nil {
			return nil, err
		}
		newObj, err := c.Assertion.Object().WalkReplace(target, replacement)
		if err != nil {
			return nil, err
		}
		if newPred.IsIdenticalTo(c.Assertion.Predicate()) && newObj.IsIdenticalTo(c.Assertion.Object()) {
			return e, nil
		}
		return NewAssertionEnvelope(EnvelopeEncodableEnvelope{newPred}, EnvelopeEncodableEnvelope{newObj}), nil
	default:
		return e, nil
	}
}

// NodesMatching returns digests of nodes matching specified criteria.
func (e *Envelope) NodesMatching(
	targetDigests map[bccomponents.Digest]struct{},
	obscureTypes []ObscureType,
) map[bccomponents.Digest]struct{} {
	result := make(map[bccomponents.Digest]struct{})
	visitor := func(envelope *Envelope, level int, edge EdgeType, state struct{}) (struct{}, bool) {
		digestMatches := true
		if targetDigests != nil {
			_, digestMatches = targetDigests[envelope.Digest()]
		}
		if !digestMatches {
			return struct{}{}, false
		}
		if len(obscureTypes) == 0 {
			result[envelope.Digest()] = struct{}{}
			return struct{}{}, false
		}
		for _, ot := range obscureTypes {
			switch ot {
			case ObscureTypeElided:
				if _, ok := envelope.envelopeCase.(*ElidedCase); ok {
					result[envelope.Digest()] = struct{}{}
				}
			case ObscureTypeEncrypted:
				if _, ok := envelope.envelopeCase.(*EncryptedCase); ok {
					result[envelope.Digest()] = struct{}{}
				}
			case ObscureTypeCompressed:
				if _, ok := envelope.envelopeCase.(*CompressedCase); ok {
					result[envelope.Digest()] = struct{}{}
				}
			}
		}
		return struct{}{}, false
	}
	e.Walk(false, struct{}{}, visitor)
	return result
}

// --- Convenience set-with-action methods ---

// ElideRemovingSetWithAction removes elements in the target set using the given
// obscure action and optional encryption key.
func (e *Envelope) ElideRemovingSetWithAction(
	target map[bccomponents.Digest]struct{},
	action ObscureAction,
	encryptKey *ObscureActionEncryptWithKey,
) *Envelope {
	return e.ElideSetWithAction(target, false, action, encryptKey)
}

// ElideRevealingSetWithAction reveals elements in the target set, obscuring
// everything else using the given action.
func (e *Envelope) ElideRevealingSetWithAction(
	target map[bccomponents.Digest]struct{},
	action ObscureAction,
	encryptKey *ObscureActionEncryptWithKey,
) *Envelope {
	return e.ElideSetWithAction(target, true, action, encryptKey)
}

// --- WalkDecrypt ---

// WalkDecrypt recursively walks the envelope and decrypts any encrypted nodes
// using the provided symmetric keys. Returns a new envelope with decrypted
// content where possible; nodes that cannot be decrypted remain unchanged.
func (e *Envelope) WalkDecrypt(keys []bccomponents.SymmetricKey) *Envelope {
	switch c := e.envelopeCase.(type) {
	case *EncryptedCase:
		for _, key := range keys {
			decrypted, err := e.DecryptSubject(key)
			if err == nil {
				return decrypted.WalkDecrypt(keys)
			}
		}
		return e
	case *NodeCase:
		newSubject := c.Subject.WalkDecrypt(keys)
		newAssertions := make([]*Envelope, len(c.Assertions))
		changed := !newSubject.IsIdenticalTo(c.Subject)
		for i, a := range c.Assertions {
			newAssertions[i] = a.WalkDecrypt(keys)
			if !newAssertions[i].IsIdenticalTo(a) {
				changed = true
			}
		}
		if !changed {
			return e
		}
		return newWithUncheckedAssertions(newSubject, newAssertions)
	case *WrappedCase:
		newEnv := c.Envelope.WalkDecrypt(keys)
		if newEnv.IsIdenticalTo(c.Envelope) {
			return e
		}
		return newEnv.Wrap()
	case *AssertionCase:
		newPred := c.Assertion.Predicate().WalkDecrypt(keys)
		newObj := c.Assertion.Object().WalkDecrypt(keys)
		if newPred.IsIdenticalTo(c.Assertion.Predicate()) && newObj.IsIdenticalTo(c.Assertion.Object()) {
			return e
		}
		return NewAssertionEnvelope(EnvelopeEncodableEnvelope{newPred}, EnvelopeEncodableEnvelope{newObj})
	default:
		return e
	}
}

// --- WalkDecompress ---

// WalkDecompress recursively walks the envelope and decompresses compressed
// nodes. If targetDigests is non-nil, only nodes whose digest is in the set
// are decompressed. If nil, all compressed nodes are decompressed.
func (e *Envelope) WalkDecompress(targetDigests map[bccomponents.Digest]struct{}) *Envelope {
	switch c := e.envelopeCase.(type) {
	case *CompressedCase:
		matchesTarget := true
		if targetDigests != nil {
			_, matchesTarget = targetDigests[e.Digest()]
		}
		if matchesTarget {
			decompressed, err := e.Decompress()
			if err == nil {
				return decompressed.WalkDecompress(targetDigests)
			}
		}
		return e
	case *NodeCase:
		newSubject := c.Subject.WalkDecompress(targetDigests)
		newAssertions := make([]*Envelope, len(c.Assertions))
		changed := !newSubject.IsIdenticalTo(c.Subject)
		for i, a := range c.Assertions {
			newAssertions[i] = a.WalkDecompress(targetDigests)
			if !newAssertions[i].IsIdenticalTo(a) {
				changed = true
			}
		}
		if !changed {
			return e
		}
		return newWithUncheckedAssertions(newSubject, newAssertions)
	case *WrappedCase:
		newEnv := c.Envelope.WalkDecompress(targetDigests)
		if newEnv.IsIdenticalTo(c.Envelope) {
			return e
		}
		return newEnv.Wrap()
	case *AssertionCase:
		newPred := c.Assertion.Predicate().WalkDecompress(targetDigests)
		newObj := c.Assertion.Object().WalkDecompress(targetDigests)
		if newPred.IsIdenticalTo(c.Assertion.Predicate()) && newObj.IsIdenticalTo(c.Assertion.Object()) {
			return e
		}
		return NewAssertionEnvelope(EnvelopeEncodableEnvelope{newPred}, EnvelopeEncodableEnvelope{newObj})
	default:
		return e
	}
}

// --- Helpers ---

func digestSet(d bccomponents.Digest) map[bccomponents.Digest]struct{} {
	return map[bccomponents.Digest]struct{}{d: {}}
}

func digestSetFromArray(providers []DigestProvider) map[bccomponents.Digest]struct{} {
	result := make(map[bccomponents.Digest]struct{}, len(providers))
	for _, p := range providers {
		result[p.Digest()] = struct{}{}
	}
	return result
}

File: python/bc-envelope/src/bc_envelope/_elide.py
"""Elision, encryption, compression (obscuration) of envelope elements.

Supports selective disclosure by obscuring parts of an envelope while
maintaining the integrity of the digest tree.
"""

from __future__ import annotations

from enum import Enum, auto
from typing import TYPE_CHECKING

from bc_components import Digest, SymmetricKey

from ._envelope_case import CaseType

if TYPE_CHECKING:
    from ._envelope import Envelope


# ---------------------------------------------------------------------------
# ObscureType — identifies how an element has been obscured
# ---------------------------------------------------------------------------


class ObscureType(Enum):
    """Identifies the kind of obscuration applied to an envelope element."""

    ELIDED = auto()
    ENCRYPTED = auto()
    COMPRESSED = auto()


# ---------------------------------------------------------------------------
# ObscureAction — what to do when obscuring an element
# ---------------------------------------------------------------------------


class _ObscureKind(Enum):
    """Internal discriminant for ObscureAction."""
    ELIDE = auto()
    ENCRYPT = auto()
    COMPRESS = auto()


class ObscureAction:
    """Action to perform when obscuring an envelope element.

    Use the class-level factory helpers:

    * ``ObscureAction.elide()``
    * ``ObscureAction.encrypt(key)``
    * ``ObscureAction.compress()``
    """

    __slots__ = ("_kind", "_key")

    def __init__(self, kind: _ObscureKind, key: SymmetricKey | None = None) -> None:
        self._kind = kind
        self._key = key

    # --- Factories ---------------------------------------------------------

    @staticmethod
    def elide() -> ObscureAction:
        return ObscureAction(_ObscureKind.ELIDE)

    @staticmethod
    def encrypt(key: SymmetricKey) -> ObscureAction:
        return ObscureAction(_ObscureKind.ENCRYPT, key)

    @staticmethod
    def compress() -> ObscureAction:
        return ObscureAction(_ObscureKind.COMPRESS)

    # --- Properties --------------------------------------------------------

    @property
    def kind(self) -> _ObscureKind:
        return self._kind

    @property
    def key(self) -> SymmetricKey | None:
        return self._key


# ---------------------------------------------------------------------------
# Envelope elision methods — attached to Envelope via __init__.py
# ---------------------------------------------------------------------------


def elide(self: Envelope) -> Envelope:
    """Return the elided variant of this envelope (digest only).

    Returns the same envelope if already elided.
    """
    from ._envelope import Envelope as Env

    if self.case_type == CaseType.ELIDED:
        return self
    return Env.new_elided(self.digest())


def elide_removing_set_with_action(
    self: Envelope,
    target: set[Digest],
    action: ObscureAction,
) -> Envelope:
    """Obscure elements whose digests are in *target*."""
    return _elide_set_with_action(self, target, False, action)


def elide_removing_set(self: Envelope, target: set[Digest]) -> Envelope:
    """Elide elements whose digests are in *target*."""
    return _elide_set(self, target, False)


def elide_removing_array_with_action(
    self: Envelope,
    target: list,
    action: ObscureAction,
) -> Envelope:
    """Obscure elements matching digest providers in *target*."""
    return _elide_array_with_action(self, target, False, action)


def elide_removing_array(self: Envelope, target: list) -> Envelope:
    """Elide elements matching digest providers in *target*."""
    return _elide_array(self, target, False)


def elide_removing_target_with_action(
    self: Envelope,
    target,
    action: ObscureAction,
) -> Envelope:
    """Obscure a single element identified by a digest provider."""
    return _elide_target_with_action(self, target, False, action)


def elide_removing_target(self: Envelope, target) -> Envelope:
    """Elide a single element identified by a digest provider."""
    return _elide_target(self, target, False)


def elide_revealing_set_with_action(
    self: Envelope,
    target: set[Digest],
    action: ObscureAction,
) -> Envelope:
    """Obscure all elements *except* those whose digests are in *target*."""
    return _elide_set_with_action(self, target, True, action)


def elide_revealing_set(self: Envelope, target: set[Digest]) -> Envelope:
    """Elide all elements *except* those whose digests are in *target*."""
    return _elide_set(self, target, True)


def elide_revealing_array_with_action(
    self: Envelope,
    target: list,
    action: ObscureAction,
) -> Envelope:
    """Obscure all elements *except* those matching digest providers."""
    return _elide_array_with_action(self, target, True, action)


def elide_revealing_array(self: Envelope, target: list) -> Envelope:
    """Elide all elements *except* those matching digest providers."""
    return _elide_array(self, target, True)


def elide_revealing_target_with_action(
    self: Envelope,
    target,
    action: ObscureAction,
) -> Envelope:
    """Obscure all elements *except* a single digest provider."""
    return _elide_target_with_action(self, target, True, action)


def elide_revealing_target(self: Envelope, target) -> Envelope:
    """Elide all elements *except* a single digest provider."""
    return _elide_target(self, target, True)


# ---------------------------------------------------------------------------
# Core dispatch
# ---------------------------------------------------------------------------

# Truth table:
#   target_matches   is_revealing   -> obscure?
#   False            False          -> False
#   False            True           -> True
#   True             False          -> True
#   True             True           -> False
# i.e. obscure = (target_matches != is_revealing)


def _elide_set_with_action(
    envelope: Envelope,
    target: set[Digest],
    is_revealing: bool,
    action: ObscureAction,
) -> Envelope:
    """Core recursive elision engine."""
    from ._assertion import Assertion
    from ._envelope import Envelope as Env

    self_digest = envelope.digest()

    if (self_digest in target) != is_revealing:
        # This element should be obscured
        if action.kind == _ObscureKind.ELIDE:
            return elide(envelope)
        elif action.kind == _ObscureKind.ENCRYPT:
            assert action.key is not None
            message = action.key.encrypt_with_digest(
                envelope.tagged_cbor().to_cbor_data(),
                self_digest,
            )
            return Env.new_with_encrypted(message)
        elif action.kind == _ObscureKind.COMPRESS:
            return envelope.compress()
        else:
            raise ValueError(f"Unknown obscure action: {action.kind}")

    case = envelope.case
    ct = case.case_type

    if ct == CaseType.ASSERTION:
        predicate = _elide_set_with_action(
            case.assertion.predicate, target, is_revealing, action
        )
        obj = _elide_set_with_action(
            case.assertion.object, target, is_revealing, action
        )
        elided_assertion = Assertion(predicate, obj)
        assert elided_assertion == case.assertion
        return Env.new_with_assertion(elided_assertion)

    if ct == CaseType.NODE:
        elided_subject = _elide_set_with_action(
            case.subject, target, is_revealing, action
        )
        assert elided_subject.digest() == case.subject.digest()
        elided_assertions = []
        for a in case.assertions:
            ea = _elide_set_with_action(a, target, is_revealing, action)
            assert ea.digest() == a.digest()
            elided_assertions.append(ea)
        return Env.new_with_unchecked_assertions(
            elided_subject, elided_assertions
        )

    if ct == CaseType.WRAPPED:
        elided_envelope = _elide_set_with_action(
            case.envelope, target, is_revealing, action
        )
        assert elided_envelope.digest() == case.envelope.digest()
        return Env.new_wrapped(elided_envelope)

    return envelope


def _elide_set(
    envelope: Envelope,
    target: set[Digest],
    is_revealing: bool,
) -> Envelope:
    return _elide_set_with_action(
        envelope, target, is_revealing, ObscureAction.elide()
    )


def _elide_array_with_action(
    envelope: Envelope,
    target: list,
    is_revealing: bool,
    action: ObscureAction,
) -> Envelope:
    digest_set = {item.digest() for item in target}
    return _elide_set_with_action(envelope, digest_set, is_revealing, action)


def _elide_array(
    envelope: Envelope,
    target: list,
    is_revealing: bool,
) -> Envelope:
    return _elide_array_with_action(
        envelope, target, is_revealing, ObscureAction.elide()
    )


def _elide_target_with_action(
    envelope: Envelope,
    target,
    is_revealing: bool,
    action: ObscureAction,
) -> Envelope:
    return _elide_array_with_action(envelope, [target], is_revealing, action)


def _elide_target(
    envelope: Envelope,
    target,
    is_revealing: bool,
) -> Envelope:
    return _elide_target_with_action(
        envelope, target, is_revealing, ObscureAction.elide()
    )


# ---------------------------------------------------------------------------
# Unelide
# ---------------------------------------------------------------------------


def unelide(self: Envelope, original: Envelope) -> Envelope:
    """Restore from *original* if digests match.

    Raises
    ------
    InvalidDigest
        If the digests do not match.
    """
    from ._error import InvalidDigest

    if self.digest() == original.digest():
        return original
    raise InvalidDigest()


# ---------------------------------------------------------------------------
# nodes_matching
# ---------------------------------------------------------------------------


def nodes_matching(
    self: Envelope,
    target_digests: set[Digest] | None,
    obscure_types: list[ObscureType],
) -> set[Digest]:
    """Return digests of nodes matching the given criteria."""
    from ._walk import EdgeType

    result: set[Digest] = set()

    def visitor(
        env: Envelope, _level: int, _edge: EdgeType, state: None
    ) -> tuple[None, bool]:
        digest_matches = (
            target_digests is None or env.digest() in target_digests
        )
        if not digest_matches:
            return (None, False)

        if not obscure_types:
            result.add(env.digest())
            return (None, False)

        ct = env.case_type
        for ot in obscure_types:
            if ot == ObscureType.ELIDED and ct == CaseType.ELIDED:
                result.add(env.digest())
                break
            if ot == ObscureType.ENCRYPTED and ct == CaseType.ENCRYPTED:
                result.add(env.digest())
                break
            if ot == ObscureType.COMPRESSED and ct == CaseType.COMPRESSED:
                result.add(env.digest())
                break

        return (None, False)

    self.walk(False, None, visitor)
    return result


# ---------------------------------------------------------------------------
# walk_unelide
# ---------------------------------------------------------------------------


def walk_unelide(self: Envelope, envelopes: list[Envelope]) -> Envelope:
    """Restore elided nodes using the provided list of envelopes."""
    envelope_map: dict[Digest, Envelope] = {}
    for env in envelopes:
        envelope_map[env.digest()] = env
    return _walk_unelide_with_map(self, envelope_map)


def _walk_unelide_with_map(
    envelope: Envelope,
    envelope_map: dict[Digest, Envelope],
) -> Envelope:
    from ._assertion import Assertion
    from ._envelope import Envelope as Env

    case = envelope.case
    ct = case.case_type

    if ct == CaseType.ELIDED:
        replacement = envelope_map.get(envelope.digest())
        if replacement is not None:
            return replacement
        return envelope

    if ct == CaseType.NODE:
        new_subject = _walk_unelide_with_map(case.subject, envelope_map)
        new_assertions = [
            _walk_unelide_with_map(a, envelope_map) for a in case.assertions
        ]
        if new_subject.is_identical_to(case.subject) and all(
            na.is_identical_to(oa)
            for na, oa in zip(new_assertions, case.assertions)
        ):
            return envelope
        return Env.new_with_unchecked_assertions(new_subject, new_assertions)

    if ct == CaseType.WRAPPED:
        new_inner = _walk_unelide_with_map(case.envelope, envelope_map)
        if new_inner.is_identical_to(case.envelope):
            return envelope
        return new_inner.wrap()

    if ct == CaseType.ASSERTION:
        new_pred = _walk_unelide_with_map(
            case.assertion.predicate, envelope_map
        )
        new_obj = _walk_unelide_with_map(case.assertion.object, envelope_map)
        if new_pred.is_identical_to(
            case.assertion.predicate
        ) and new_obj.is_identical_to(case.assertion.object):
            return envelope
        return Env.new_assertion(new_pred, new_obj)

    return envelope


# ---------------------------------------------------------------------------
# walk_replace
# ---------------------------------------------------------------------------


def walk_replace(
    self: Envelope,
    target: set[Digest],
    replacement: Envelope,
) -> Envelope:
    """Replace nodes whose digests are in *target* with *replacement*.

    Raises
    ------
    InvalidFormat
        If an assertion would be replaced by a non-assertion,
        non-obscured envelope.
    """
    return _walk_replace_impl(self, target, replacement)


def _walk_replace_impl(
    envelope: Envelope,
    target: set[Digest],
    replacement: Envelope,
) -> Envelope:
    from ._envelope import Envelope as Env

    if envelope.digest() in target:
        return replacement

    case = envelope.case
    ct = case.case_type

    if ct == CaseType.NODE:
        new_subject = _walk_replace_impl(case.subject, target, replacement)
        new_assertions = [
            _walk_replace_impl(a, target, replacement) for a in case.assertions
        ]
        if new_subject.is_identical_to(case.subject) and all(
            na.is_identical_to(oa)
            for na, oa in zip(new_assertions, case.assertions)
        ):
            return envelope
        # Use validated constructor to check assertion validity
        return Env.new_with_assertions(new_subject, new_assertions)

    if ct == CaseType.WRAPPED:
        new_inner = _walk_replace_impl(case.envelope, target, replacement)
        if new_inner.is_identical_to(case.envelope):
            return envelope
        return new_inner.wrap()

    if ct == CaseType.ASSERTION:
        new_pred = _walk_replace_impl(
            case.assertion.predicate, target, replacement
        )
        new_obj = _walk_replace_impl(
            case.assertion.object, target, replacement
        )
        if new_pred.is_identical_to(
            case.assertion.predicate
        ) and new_obj.is_identical_to(case.assertion.object):
            return envelope
        return Env.new_assertion(new_pred, new_obj)

    return envelope


# ---------------------------------------------------------------------------
# walk_decrypt
# ---------------------------------------------------------------------------


def walk_decrypt(self: Envelope, keys: list[SymmetricKey]) -> Envelope:
    """Recursively decrypt encrypted nodes using the provided keys."""
    return _walk_decrypt_impl(self, keys)


def _walk_decrypt_impl(
    envelope: Envelope,
    keys: list[SymmetricKey],
) -> Envelope:
    from ._envelope import Envelope as Env

    case = envelope.case
    ct = case.case_type

    if ct == CaseType.ENCRYPTED:
        for key in keys:
            try:
                decrypted = envelope.decrypt_subject(key)
                return _walk_decrypt_impl(decrypted, keys)
            except Exception:
                continue
        return envelope

    if ct == CaseType.NODE:
        new_subject = _walk_decrypt_impl(case.subject, keys)
        new_assertions = [
            _walk_decrypt_impl(a, keys) for a in case.assertions
        ]
        if new_subject.is_identical_to(case.subject) and all(
            na.is_identical_to(oa)
            for na, oa in zip(new_assertions, case.assertions)
        ):
            return envelope
        return Env.new_with_unchecked_assertions(new_subject, new_assertions)

    if ct == CaseType.WRAPPED:
        new_inner = _walk_decrypt_impl(case.envelope, keys)
        if new_inner.is_identical_to(case.envelope):
            return envelope
        return new_inner.wrap()

    if ct == CaseType.ASSERTION:
        new_pred = _walk_decrypt_impl(case.assertion.predicate, keys)
        new_obj = _walk_decrypt_impl(case.assertion.object, keys)
        if new_pred.is_identical_to(
            case.assertion.predicate
        ) and new_obj.is_identical_to(case.assertion.object):
            return envelope
        return Env.new_assertion(new_pred, new_obj)

    return envelope


# ---------------------------------------------------------------------------
# walk_decompress
# ---------------------------------------------------------------------------


def walk_decompress(
    self: Envelope,
    target_digests: set[Digest] | None = None,
) -> Envelope:
    """Recursively decompress compressed nodes.

    If *target_digests* is provided, only decompress nodes whose digests
    are in the set.
    """
    return _walk_decompress_impl(self, target_digests)


def _walk_decompress_impl(
    envelope: Envelope,
    target_digests: set[Digest] | None,
) -> Envelope:
    from ._envelope import Envelope as Env

    case = envelope.case
    ct = case.case_type

    if ct == CaseType.COMPRESSED:
        matches = (
            target_digests is None or envelope.digest() in target_digests
        )
        if matches:
            try:
                decompressed = envelope.decompress()
                return _walk_decompress_impl(decompressed, target_digests)
            except Exception:
                pass
        return envelope

    if ct == CaseType.NODE:
        new_subject = _walk_decompress_impl(case.subject, target_digests)
        new_assertions = [
            _walk_decompress_impl(a, target_digests) for a in case.assertions
        ]
        if new_subject.is_identical_to(case.subject) and all(
            na.is_identical_to(oa)
            for na, oa in zip(new_assertions, case.assertions)
        ):
            return envelope
        return Env.new_with_unchecked_assertions(new_subject, new_assertions)

    if ct == CaseType.WRAPPED:
        new_inner = _walk_decompress_impl(case.envelope, target_digests)
        if new_inner.is_identical_to(case.envelope):
            return envelope
        return new_inner.wrap()

    if ct == CaseType.ASSERTION:
        new_pred = _walk_decompress_impl(
            case.assertion.predicate, target_digests
        )
        new_obj = _walk_decompress_impl(
            case.assertion.object, target_digests
        )
        if new_pred.is_identical_to(
            case.assertion.predicate
        ) and new_obj.is_identical_to(case.assertion.object):
            return envelope
        return Env.new_assertion(new_pred, new_obj)

    return envelope
